### DISCUSSION TITLE
Problems with Transitions

### DISCUSSION THREAD

**Discussion by IcemarkUK** [axmolengine/axmol#1663] [Q&A] [answered]
I'm having a few issues with transitions.

Firstly, TransitionCrossFade is not fading between two scenes. It seems to fade between a blank scene and the one I'm leaving. The scene I'm entering appears at the end of the duration. If I s

Here is my trigger of the transition.
```
        Director* director = Director::getInstance();
        director->pushScene(TransitionCrossFade::create(defaultDuration, incomming) );
```

https://github.com/axmolengine/axmol/assets/8301416/a8cd49b3-db95-4dfa-8335-9c03d0035842

One of my Scenes has a ScrollView with a TMXTiledMap inside it. If I CrossFade to that scene it crashes in the following code. Basically getVisitingCamera is null.

If I change to another transition, it works fine.

```
void FastTMXLayer::draw(Renderer* renderer, const Mat4& transform, uint32_t flags)
{
    updateTotalQuads();

    auto cam = Camera::getVisitingCamera();
    if (flags != 0 || _dirty || _quadsDirty || !_cameraPositionDirty.fuzzyEquals(cam->getPosition(), _tileSet->_tileSize.x) ||
        _cameraZoomDirty != cam->getZoom())
    {
        _cameraPositionDirty = cam->getPosition();
```

And finally - the slide transitions are tearing really badly. Couldn't capture in video as the seem fine in the recording!

**Comment by rh101** [comment]
> Firstly, TransitionCrossFade is not fading between two scenes. It seems to fade between a blank scene and the one I'm leaving. The scene I'm entering appears at the end of the duration.

It's definitely different from the way it worked in cocos2d-x, and it seems that no one noticed.  The cpp-tests project also shows the issue, where it fades to a black screen before popping up the new scene, which is not correct.

**Reply by rh101** [reply, marked as answer]
@IcemarkUK Fixed in #1664

**Reply by IcemarkUK** [reply]
Perfect, thank you that works.

**Comment by rh101** [comment]
> One of my Scenes has a ScrollView with a TMXTiledMap inside it. If I CrossFade to that scene it crashes in the following code. Basically getVisitingCamera is null.

> And finally - the slide transitions are tearing really badly. Couldn't capture in video as the seem fine in the recording!

Any chance you could make a simple test project, or even just post the code, which reproduces each of these issues?

**Reply by IcemarkUK** [reply]
I'll look at the tearing. See if I can create something.

With the cross fade in to TMXTiledMap crash, the crash is inside 
Director -> drawScene -> setNextScene ->
    TransitionCrossFade::onEnter ...

Interestingly, I noticed this in the code for drawScene

```
   /* to avoid flickr, nextScene MUST be here: after tick and before draw.
     * FIXME: Which bug is this one. It seems that it can't be reproduced with v0.9
     */
    if (_nextScene)
    {
        setNextScene();
    }

```

I suspect the Node isn't being set up properly somehow by the time the first it's visited. I'll dig further (I can always change/remove the transition if it's holding me up)

But should you have the time or inclination, my repos is open, [here is the branch](https://github.com/ChilliHugger/The-Lords-Of-Midnight/tree/AXMOL-port)...

**Reply by rh101** [reply]
> But should you have the time or inclination, my repos is open, [here is the branch](https://github.com/ChilliHugger/The-Lords-Of-Midnight/tree/AXMOL-port)...

Excellent! I'll definitely check it out.

**Reply by rh101** [reply]
> Interestingly, I noticed this in the code for drawScene
> 
> ```
>    /* to avoid flickr, nextScene MUST be here: after tick and before draw.
>      * FIXME: Which bug is this one. It seems that it can't be reproduced with v0.9
>      */
>     if (_nextScene)
>     {
>         setNextScene();
>     }
> ```

That's nothing new to Axmol, as it's been inherited from Cocos2d-x: https://github.com/cocos2d/cocos2d-x/blob/76903dee64046c7bfdba50790be283484b4be271/cocos/base/CCDirector.cpp#L273

**Reply by IcemarkUK** [reply]
@rh101 
Changing the following
https://github.com/axmolengine/axmol/blob/dad167bb2e3a3b398828b0b8228c932a3759f945/core/2d/FastTMXLayer.cpp#L131-L138

To add a null check at L136, fixes the issue. The transition isn't working properly, but at least it doesn't crash.
```
if (cam == nullptr) return;
```

**Reply by IcemarkUK** [reply]
This is essentially because setNextScene is called before the cameras are set up in renderScene. And this happens with all transitions. 

TransitionCrossFade seems to be the only transition that calls `_inScene->visit()` inside `onEnter` which triggers a `Draw` - which it does in order to create the render textures. The OnEnter is triggered inside setNextScene.

**Reply by rh101** [reply]
Interesting find, and I wonder when it changed between Cocos2d-x and Axmol, and the reason why.  

Please submit a PR with the modification that fixes the issue.

**Reply by rh101** [reply]
The issue seems to have been introduced in this PR, which was to fix a different issue: https://github.com/axmolengine/axmol/pull/1134

Adding that null check is required to fix the issue you're experiencing, and similar checks are used in draw code of other nodes that access the camera at that point.

Does the fact that it does not draw anything affect the transition, in terms of showing the correct output from the tiled map?

**Reply by IcemarkUK** [reply]
Adding the null check stops the render texture from capturing the initial scene, so the transition doesn't cross fade correctly. So it's not a fix. 
I don't think that #1134 fix broke this, as the code previously still accessed the camera to get the position. Quickly swapping out the function for the Cocos2d-x one still crashes. I suspect it has something to do with the transitions triggering setNextScene and maybe that didn't happen before. I'll follow that process when I'm back home in a couple of days and can run the Cocos build.

**Comment by rh101** [comment]
@IcemarkUK The entire issue with this is that the camera isn't being set prior to processing all internal nodes.  I'm not sure why it would have worked in Cocos2d-x. since none of the transition code seems to have changed between cocos2d-x and axmol, including when it switches to the next scene.

There is a potential fix that, to my knowledge, will not affect anything but this specific issue.  A few observations related to it:

* `Scene:visit()` is unused outside of attempting to render a scene to a render texture.
* `Node::visit()` ~~is both virtual and final, which is strange, since it should be one or the other, and I don't know the reason why it has the `final` specifier.~~ It needs to be virtual because it is overloaded.

My proposal here is that the `final` specifier is removed from `Node::visit()`, yet keep it as a virtual method.  In this case, it actually helps fix this issue, since we can override it in `Scene::visit()`, and simply implement the following:

```
void Scene::visit()
{
    const auto eyeTransform = Mat4::IDENTITY;

    for (const auto& camera : getCameras())
    {
        if (!camera->isVisible())
            continue;

        Camera::_visitingCamera = camera;

        camera->setAdditionalTransform(eyeTransform.getInversed());
        _director->pushMatrix(MATRIX_STACK_TYPE::MATRIX_STACK_PROJECTION);
        _director->loadMatrix(MATRIX_STACK_TYPE::MATRIX_STACK_PROJECTION,
                              Camera::_visitingCamera->getViewProjectionMatrix());

        camera->apply();
        // clear background with max depth
        camera->clearBackground();
        // visit the scene
        Node::visit();

        _director->popMatrix(MATRIX_STACK_TYPE::MATRIX_STACK_PROJECTION);
    }

    Camera::_visitingCamera = nullptr;
}
```
That effectively fixes the crash, and ensures the transitions work correctly, since the camera is now valid for all child nodes of the scene,. Transitions use something like the following calls to output to the RenderTexture:

```
void TransitionCrossFade::onEnter()
{
...
    inTexture->begin();
    _inScene->visit();
    inTexture->end();
...
}
```

@halx99 Is this a reasonable patch? `Scene::visit()` is never called at all except when attempting to render it to a RenderTexture, which is what occurs in the scene transition code.  This change shouldn't impact anything else in the engine.

**Reply by rh101** [reply]
Just some more info.  The scene is normally rendered from the glview implementation, for example:

```
void GLView::renderScene(Scene* scene, Renderer* renderer)
{
    AXASSERT(scene, "Invalid Scene");
    AXASSERT(renderer, "Invalid Renderer");

    scene->render(renderer, Mat4::IDENTITY, nullptr);
}
```

Inside `Scene::render(...)`, this method is called:

`visit(renderer, transform, 0);`

So, it never uses the `Scene::visit()` method at all.

**Comment by rh101** [comment]
@IcemarkUK PR #1676 is up with the proposed fix for the crash issue.

**Reply by IcemarkUK** [reply]
Many thanks!
